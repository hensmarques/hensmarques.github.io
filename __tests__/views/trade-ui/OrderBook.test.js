import React from 'react';
import { createMockStore } from 'redux-test-utils';
import OrderBook from '../../../src/views/trade-ui/OrderBook';
import types from '../../../src/actions/types';

const mockState = {
  selectedInstrument: {
    id: 'BTCUSD',
    base: 'BTC',
    quote: 'USD',
    quoteDecimals: 5,
    quantityDecimals: 4,
    type: 'crypto',
    quantityIncrement: '0.0001',
    fees: {},
  },
  orderbook: {
    asks: [
      { quantity: '0.07', number_of_orders: null, price: 9415.035 },
      { quantity: '0.01', number_of_orders: null, price: 9417.53 },
    ],
    bids: [
      { quantity: '0.189', number_of_orders: null, price: 9410.84 },
      { quantity: '0.4', number_of_orders: null, price: 9413.74 },
    ],
  },
};

// Mock necessary for centerBookScroll function
document.querySelector = jest.fn().mockImplementation(() => ({
  clientHeight: 100,
}));

describe('<OrderBook/>', () => {
  const store = createMockStore(mockState);
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithStore(<OrderBook />, store);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('loads data and centers book scroll on mount', () => {
    expect(
      store.isActionDispatched({
        type: types.app.loadL2Data,
        instrument: mockState.selectedInstrument.id,
      }),
    ).toBe(true);
    const instance = wrapper.find('OrderBook').instance();
    jest.spyOn(instance, 'centerBookScroll');
    instance.componentDidMount();
    expect(instance.centerBookScroll).toHaveBeenCalledTimes(1);
  });

  it('renders the order book with L2 data', () => {
    expect(
      wrapper.containsMatchingElement(
        <div className="order-book-header">
          <span className="order-book-header-price">
            {`Price ${mockState.selectedInstrument.quote}`}
          </span>
          <span className="order-book-header-quantity">
            {`Quantity ${mockState.selectedInstrument.base}`}
          </span>
        </div>,
      ),
    ).toBe(true);
    expect(
      wrapper.containsMatchingElement(
        <div className="order-book-spread">
          {`Spread ${(
            mockState.orderbook.asks[0].price - mockState.orderbook.bids[0].price
          ).toFixed(mockState.selectedInstrument.quoteDecimals)} ${
            mockState.selectedInstrument.quote
          }`}
        </div>,
      ),
    ).toBe(true);

    expect(wrapper.find('.order-book-asks')).toMatchSnapshot();
    expect(wrapper.find('.order-book-bids')).toMatchSnapshot();
  });
});
