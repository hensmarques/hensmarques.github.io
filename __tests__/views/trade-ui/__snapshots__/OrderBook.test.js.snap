// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBook/> renders the order book with L2 data 1`] = `
<div
  className="order-book-asks"
>
  <div
    className="order-book-row"
    key="0"
  >
    <span
      className="order-book-row-price"
    >
      <span>
        <span
          className="highlight"
        >
          9415
        </span>
        <span
          className="highlight"
        >
          .
          035
        </span>
        <span>
          00000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-quantity"
    >
      <span>
        <span
          className=""
        >
          0
        </span>
        <span
          className="highlight"
        >
          .
          07
        </span>
        <span>
          000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-open-orders"
    />
  </div>
  <div
    className="order-book-row"
    key="1"
  >
    <span
      className="order-book-row-price"
    >
      <span>
        <span
          className="highlight"
        >
          9417
        </span>
        <span
          className="highlight"
        >
          .
          53
        </span>
        <span>
          000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-quantity"
    >
      <span>
        <span
          className=""
        >
          0
        </span>
        <span
          className="highlight"
        >
          .
          01
        </span>
        <span>
          000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-open-orders"
    />
  </div>
</div>
`;

exports[`<OrderBook/> renders the order book with L2 data 2`] = `
<div
  className="order-book-bids"
>
  <div
    className="order-book-row"
    key="0"
  >
    <span
      className="order-book-row-price"
    >
      <span>
        <span
          className="highlight"
        >
          9410
        </span>
        <span
          className="highlight"
        >
          .
          84
        </span>
        <span>
          000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-quantity"
    >
      <span>
        <span
          className=""
        >
          0
        </span>
        <span
          className="highlight"
        >
          .
          189
        </span>
        <span>
          00000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-open-orders"
    />
  </div>
  <div
    className="order-book-row"
    key="1"
  >
    <span
      className="order-book-row-price"
    >
      <span>
        <span
          className="highlight"
        >
          9413
        </span>
        <span
          className="highlight"
        >
          .
          74
        </span>
        <span>
          000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-quantity"
    >
      <span>
        <span
          className=""
        >
          0
        </span>
        <span
          className="highlight"
        >
          .
          4
        </span>
        <span>
          0000000
        </span>
      </span>
    </span>
    <span
      className="order-book-row-open-orders"
    />
  </div>
</div>
`;
